plat/common/include/uk/plat/common/lcpu.h

```diff
@@ -0,0 +1,353 @@
+	  : NULL)
+
+#define lcpu_lcpuidx_list_foreach(list, n, i, lcpu)			\
+	if ((n) == 0) {							\
+		UK_ASSERT(!(list));					\
+		(n) = ukplat_lcpu_count();				\
+	} else								\
+		UK_ASSERT((n) <= ukplat_lcpu_count());			\
+	for ((i) = 0, (lcpu) = _lcpu_lcpuidx_list_entry(list, i, n);	\
+	     (i) < (n);							\
+	     (i)++, (lcpu) = _lcpu_lcpuidx_list_entry(list, i, n))
+
+/**
+ * Return the LCPU structure for the CPU executing this function
+ */
+struct lcpu *lcpu_get_current(void);
+
+/**
+ * Initialize a logical CPU. The function must be executed on the CPU
+ * represented by the LCPU as early as possible after startup.
+ *
+ * @param this_lcpu pointer to the LCPU structure representing the CPU
+ *    executing this function
+ * @return 0 on success, -errno otherwise
+ */
+int lcpu_init(struct lcpu *this_lcpu);
+
+/**
+ * Default entry function for secondary CPUs. This will call lcpu_init() and
+ * enter a low-power state to wait for interrupts and calls to
+ * ukplat_lcpu_run() destined to this CPU.
+ *
+ * NOTE: The function does not return
+ *
+ * @param this_lcpu pointer to the LCPU structure representing the CPU
+ *    executing this function
+ */
+void __noreturn lcpu_entry_default(struct lcpu *this_lcpu);
+
+#ifdef CONFIG_HAVE_SMP
+/**
+ * Initialize multi-processor functions. Must only be executed once on the
+ * bootstrap processor (BSP)
+ *
+ * @param arg an optional parameter from the boot code that is passed to the
+ *    architectural initialization
+ * @return 0 on success, -errno otherwise
+ */
+int lcpu_mp_init(void *arg);
+#endif /* CONFIG_HAVE_SMP */
+
+/*
+ * Definitions that must be satisfied by the architectural implementation.
+ * DO NOT CALL DIRECTLY. Use the ukplat_* and lcpu_* non-architectural versions.
+ */
+
+/**
+ * Return the hardware ID of the CPU executing this function. Must be able to
+ * return the ID of the bootstrap processor without initialization of the MP
+ * functions.
+ */
+__lcpuid lcpu_arch_id(void);
+
+/**
+ * Initialize the architectural part of the LCPU. The function must be
+ * executed on the CPU represented by the LCPU as part of lcpu_init().
+ *
+ * @param this_lcpu pointer to the LCPU structure representing the CPU
+ *    executing this function
+ * @return 0 on success, -errno otherwise
+ */
+int lcpu_arch_init(struct lcpu *this_lcpu);
+
+#ifdef CONFIG_HAVE_SMP
+/**
+ * Initialize the architectural part of the multi-processor functions. This
+ * should perform CPU discovery and call lcpu_alloc() for each discovered CPU.
+ * The bootstrap processor is already allocated and must not be added.
+ *
+ * @param bsp_cpu_id id of the bootstrap processor
+ * @param arg an optional parameter from the boot code. Can be NULL
+ *
+ * @return 0 on success, -errno otherwise
+ */
+int lcpu_arch_mp_init(__lcpuid bsp_cpu_id, void *arg);
+
+/**
+ * Start the given logical CPU. The CPU should execute the entry function
+ * with the supplied stack. The CPU will be in INIT state.
```

Comment:
Maybe "Upon return, the CPU will be in INIT state." would be more clear.

Reply:
Actually, it will be in `INIT` state when **entering** the function. Perhaps here is a good place to put more information on the expected CPU state prior call and before calling the entry function.

Reply:
> The bootcore has to run lcpu_init() and lcpu_mp_init().

I think that proves my point about the need for clarification then 😄

Reply:
This one is still pending.